Use RPC context streams to cover serde ops

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57958 [Don't Review] Enforce stream synchornization after optim.step
* #57918 [WIP] Fix distributed autograd gradient accumulation
* #57917 [WIP] Fix Raw RPC/RRef Micro-batch-based Pipelining
* #57896 Revert "Make RRefContext get devices from RPC agent when creating OwnerRRef (#57443)"
* **#57926 Use RPC context streams to cover serde ops**

Differential Revision: [D28316526](https://our.internmc.facebook.com/intern/diff/D28316526)

mrshenli committed:
Use RPC context streams to cover serde ops

[ghstack-poisoned]

mrshenli commented:
Plan to land this before branch cut.

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28316526).

mrshenli committed:
Update on "Use RPC context streams to cover serde ops"


Differential Revision: [D28316526](https://our.internmc.facebook.com/intern/diff/D28316526)

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28316526).

lw commented on torch/csrc/distributed/rpc/tensorpipe_utils.cpp:
```diff
@@ -47,6 +47,12 @@ std::tuple<tensorpipe::Message, TensorpipeWriteBuffers> tensorpipeSerialize(
   tensorpipe::Message tpMessage;
   TensorpipeWriteBuffers buffers;
 
+  // The pickler below might allocate new tensors if RPC arguments contain
+  // Tensor views. Apply stream guard here to include those Tensor allocation
+  // operations to the streams.
+  c10::MultiStreamGuard guard(
+          ctx ? ctx->getReservedStreams() : ArrayRef<Stream>({}));
```
I guess this happens within `cloneSparseTensors`? If so, it might be better to move this closer to where we invoke `cloneSparseTensors` and, perhaps, limit this guard to a narrower scope. This would put these two pieces of code, which are related, closer together, and would ensure that anyone who modifies one of them is aware of the other.

lw commented on torch/csrc/distributed/rpc/tensorpipe_utils.cpp:
```diff
@@ -230,7 +236,15 @@ std::pair<tensorpipe::Allocation, TensorpipeReadBuffers> tensorpipeAllocate(
+    TensorpipeReadBuffers&& buffers,
+    const std::shared_ptr<LazyStreamContext>& ctx) {
+
+  // I believe we don't need this today, as unlike pickling, unpickling does
+  // not insert CUDA ops. But still putting this guard here for safety and
+  // potential future changes.
```
I'm a bit conflicted about adding this "just in case". I understand the reasoning, but it gives the impression we don't know what our code does and we're shooting in the dark.

lw commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -5829,6 +5829,29 @@ def test_owner_rref_forward_synchronization3(self):
+            rpc_backend_options=options,
+        )
+
+        x = torch.randn(1000, 200).cuda(0)
+        futs = []
+        for xx in x.split(100):
+            futs.append(rpc.rpc_async(dst, RpcTest._identity, args=(xx,)))
```
I don't see how this test triggered the bug: here there's only always one stream, which is always the current one, so the device guard we added shouldn't be having any effect. Are you sure this failed before? Could we make it clearer what we're testing for?

Also, I'd like us to stop relying on subtle timing races to catch these issues, it's too unreliable. If we intend to test that one operation waits for another one, we should explicitly delay the "earlier" operation by an insanely large amount (say, one second) using `torch.cuda._sleep` so that it becomes exceedingly unlikely that the "later" operation just casually happens to synchronize correctly. You can see how I did it in the tests for CUDAFuture just below.

lw reviewed (approved):
Accepting to unblock.

mrshenli committed:
Update on "Use RPC context streams to cover serde ops"


Differential Revision: [D28316526](https://our.internmc.facebook.com/intern/diff/D28316526)

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28316526).
